### DISCUSSION TITLE
How to set html tag title from a route

### DISCUSSION THREAD

**Discussion by giacomorebonato** [TanStack/router#1056] [Q&A]
I can't see this from the examples, I am looking for a way to set HTML page title from a route definition.

```html
<!DOCTYPE html>
<html>
<head>
  <title><!-- this title --></title>
</head>
<body>
```

This is challenging because the `<head />` element is defined in the root component `__root.tsx` and other routes don't have access to it.
At the moment I am relying on rendering an hidden HTML element that I copy to the page head, before sending the code to the browser.

Which works well, until I want to find a way to apply the streaming technique and then I am lost!

**Comment by giacomorebonato** [comment]
Another option that I am considering is to extract the `__root.tsx` component into a re-usable one, where I can pass the `title` as prop.

**Comment by dohomi** [comment]
Following this diccussion, I've seen https://tanstack.com/router/v1/docs/guide/router-context#processing-accumulated-route-context but not sure how to adjust Vite config to be able to have the `<head/>` section inside of __root.tsx

**Reply by thinhbg2812** [reply]
Link is dead boss

**Comment by tannerlinsley** [comment]
I’m working on a first class solution for title tags.

**Reply by DevlinRocha** [reply]
first class support would be fantastic! i hope it is for the entire `<head>` element and not just the `<title>`, this is very important for SEO, such as defining canonical URLs, keywords, etc. for each route. thanks for all your hard work!

**Reply by moshOntong-IT** [reply]
What is the status?

**Reply by Eigilak** [reply]
❤️

**Reply by eduvhc** [reply]
Up

**Reply by thinhbg2812** [reply]
up

**Reply by ilarramendi** [reply]
Up

**Reply by nicoandreoli2000** [reply]
Updates?

**Reply by BrendanC23** [reply]
This has been implemented. [Here](https://tanstack.com/router/latest/docs/framework/react/guide/document-head-management) are the docs. Also, see the new posts at the bottom of this discussion.

**Comment by benweier** [comment]
I'd be keen for a first-class solution too. I had to adapt this solution from my previous React Router version...

For the moment I have a [`Meta` component](https://github.com/benweier/wayfarer/blob/7853b32148a16ffade31b2640ef0d14b8df34e3d/src/components/meta/meta.component.tsx) that gets all the matched routes which have a `meta` function on the route context from [`beforeload`](https://github.com/benweier/wayfarer/blob/7853b32148a16ffade31b2640ef0d14b8df34e3d/src/routes/conf/_dashboard.route.ts), and rendered as [high up in my route tree as possible](https://github.com/benweier/wayfarer/blob/7853b32148a16ffade31b2640ef0d14b8df34e3d/src/routes/core.route.tsx#L55)

The `meta` function accepts a `t` function from i18next as well as the loader data so that I can use [translations and dynamic data](https://github.com/benweier/wayfarer/blob/7853b32148a16ffade31b2640ef0d14b8df34e3d/src/routes/systems/system/system-route.meta.ts) in the meta tags if needed. I didn't want to enforce the `t` usage inside `Meta`, opting to pass it as an argument instead, and awaiting deferred data in `meta` turned out to be a terrible experience so I simply enforce that any route data used by the meta functions must never be deferred.

It's not perfect and there's actually no type-safety between my route data and the meta data argument, but it does the job pretty well - I just have to be very mindful of the loader data and manually provide the types to `meta` when required. A cool pattern I've used in another project is being able to render `Meta` components further down the route tree to effectively "override" the top level `Meta` if, for example, it needed a different `titleTemplate` pattern.

**Comment by matclayton** [comment]
@tannerlinsley any update when a first class solution might land, we're evaluating swapping to TanStack Router and this is one of the core pieces we're considering, especially when combined with Streaming responses. I noted the code/examples cover the Meta data, (except I couldn't find an example which loads in data for the tags), and wondered if this was the final API, or if it was a matter of just adding docs?

**Comment by sakibhasandev** [comment]
@tannerlinsley thanks for the wonderful library, looking forward to use the title tag feature in TanStack Router

**Comment by mutewinter** [comment]
For now, here's an example using accumulated route context:

```tsx
interface RootContext {
  getTitle?: () => string | Promise<string>;
}

function Root() {
  const matches = useMatches();

  useEffect(() => {
    // Set document title based on lowest matching route with a title
    const breadcrumbPromises = [...matches]
      .reverse()
      .map((match, index) => {
        if (!('getTitle' in match.routeContext)) {
          if (index === 0 && import.meta.env.DEV) {
            // eslint-disable-next-line no-console
            console.warn('no getTitle', match.pathname, match);
          }
          return undefined;
        }
        const { routeContext } = match;
        return routeContext.getTitle();
      })
      .filter(Boolean);
    void Promise.all(breadcrumbPromises).then((titles) => {
      document.title = titles.join(' · ');
      return titles;
    });
  }, [matches]);

  return <Outlet />;
}

export const Route = createRootRouteWithContext<RootContext>()({
  beforeLoad: () => ({ getTitle: () => 'App Name' }),
  component: Root,
});
```

```tsx
// In another route file:

export const Route = createFileRoute('/test/')({
  beforeLoad: () => ({ getTitle: () => 'Test' }),
  component: () => <Test />,
});
```

This results in title like:

- `/`: App Name
- `/test/`: Test  · App Name

And supports further nesting.

**Reply by barrenechea** [reply]
I wonder how to make it work with LazyFileRoute

**Reply by bruceharrison1984** [reply]
@barrenechea See this issue for how to make this work with Lazy routes. I just ran a quick POC and it works.

https://github.com/TanStack/router/issues/1209#issuecomment-1960652027

**Reply by barrenechea** [reply]
That was it! Thank you very much @bruceharrison1984 🚀

**Reply by thinhbg2812** [reply]
It's work for me :( 
in my case it's nested routed
```
no getTitle /order/ 
```

![CleanShot 2025-01-25 at 01 32 57](https://github.com/user-attachments/assets/11f80fac-c2cf-465b-8fa4-0a1fb5985e4a)

**Comment by giacomorebonato** [comment]
I am now relying on [react-helmet-async](https://www.npmjs.com/package/react-helmet-async) for head markup in my [fastrat](https://www.fastrat.dev/) React + Fastify starter kit.
I will update it if in the future there will be cleaner option, but I don't miss anything at the moment.

**Comment by Rendez** [comment]
I've opted for a much simpler solution relying on the `meta` property, so it's future-proof:

```tsx
const TITLE = 'Your Company';

export const Route = createRootRouteWithContext<RootRouteContext>()({
  meta: () => [
    {
      title: TITLE,
    },
  ],
  component: RootComponent,
});

function Meta({ children }: { children: ReactNode }) {
  const matches = useMatches();
  const meta = matches.at(-1)?.meta?.find((meta) => meta.title);

  useEffect(() => {
    document.title = [meta?.title, TITLE].filter(Boolean).join(' · ');
  }, [meta]);

  return children;
}

function RootComponent() {
  return (
    <Meta>
      <Outlet />
    </Meta>
  );
}
```

**Comment by bruceharrison1984** [comment]
To build on @Rendez solution, I wrote a simple hook that gathers up all the titles, and exposes them via prop. You can easily combine this with the `Meta` resource in Rendez's answer to nicely split apart all the logic from breadcrumbs.

```tsx
import { useMatches } from "@tanstack/react-router";
import {
  createContext,
  PropsWithChildren,
  useContext,
  useEffect,
  useState,
} from "react";

const BreadcrumbContext = createContext<
  | {
      breadcrumbs: string[];
      getBreadcrumbString: (separator?: string) => string;
    }
  | undefined
>(undefined);

export const BreadcrumbProvider = ({ children }: PropsWithChildren) => {
  const [breadcrumbs, setBreadcrumbs] = useState<string[]>([]);

  const matches = useMatches();

  useEffect(() => {
    const meta = matches
      .map((x) => x.meta?.find((meta) => meta.title)?.title)
      .filter((x) => x !== undefined);
    setBreadcrumbs(meta);
  }, [matches]);

  const getBreadcrumbString = (separator = " > ") => {
    if (!breadcrumbs) return "";
    return breadcrumbs?.join(separator);
  };

  return (
    <BreadcrumbContext.Provider value={{ breadcrumbs, getBreadcrumbString }}>
      {children}
    </BreadcrumbContext.Provider>
  );
};

export const useBreadcrumbs = () => {
  const breadcrumbs = useContext(BreadcrumbContext);
  if (!breadcrumbs)
    throw new Error("useBreadcrumbs must be used within BreadcrumbProvider");
  return breadcrumbs;
};

/// access breadcrumb data from any where via
/// const { breadcrumbs, getBreadcrumbString } = useBreadcrumbs();
```

Adjust your root module to be similar to the following:

```tsx
function RootComponent() {
  return (
      <BreadcrumbProvider>
        <Meta>
          <Outlet />
        </Meta>
      </BreadcrumbProvider>
  );
}
```

**Comment by gnllucena** [comment]
hey @tannerlinsley , I’ve seen the `meta` function in some examples in the tanstack/router repo:

```
import { createFileRoute } from "@tanstack/react-router";

export const Route = createFileRoute("/(identity)/sign-up")({
  meta: () => [
    {
      title: "New title",
    },
    {
      charSet: "UTF-8",
    },
    {
      name: "viewport",
      content: "width=device-width, initial-scale=1.0",
    },
  ],
});
```

but I don't see my page title changing...  is this API still experimental?

**Reply by chrism1996** [reply]
> hey @tannerlinsley , I’ve seen the `meta` function in some examples in the tanstack/router repo:
> 
> ```
> import { createFileRoute } from "@tanstack/react-router";
> 
> export const Route = createFileRoute("/(identity)/sign-up")({
>   meta: () => [
>     {
>       title: "New title",
>     },
>     {
>       charSet: "UTF-8",
>     },
>     {
>       name: "viewport",
>       content: "width=device-width, initial-scale=1.0",
>     },
>   ],
> });
> ```
> 
> but I don't see my page title changing... is this API still experimental?

Same. Trying to build a reusable Seo component to update title, description etc. I have seen the meta property in createFileRoute but i dont see it in the docs, is that experimental @tannerlinsley  ?

**Reply by vKongv** [reply]
Seen that too, seems to be under "start" documentation based on the URL. Doesn't work in router yet.

**Reply by rahulvils** [reply]
If you have `index.html` that contains `title` tag then remove it and add `HeadContent` in the __root file

**Reply by maelp** [reply]
Same here... I don't have <title> in my index.html, and I've added HeadContent and Scripts

**Comment by fitimbytyqi** [comment]
Has this been fixed ?

I just tried setting a meta attribute inside createFileRoute and it seemd to work!

```
export const Route = createFileRoute("/")({
  component: Home,
  loader: async () => await getCount(),
  meta: () => [
    {
      title: "This is working!",
    },
  ],
});
```

**Reply by nemeseri** [reply]
Can you give me some information on the version you use? It's still not working for me.

**Reply by BrendanC23** [reply]
I think you need to be using TanStack Start for the `meta` option to work out of the box. If you're just using the router, it needs a bit more code. See [this discussion](https://github.com/TanStack/router/discussions/1056#discussioncomment-10275259).

**Reply by fitimbytyqi** [reply]
@nemeseri I am sorry, as mentioned, my example comes from Start.

**Reply by thinhbg2812** [reply]
i dont see any option for meta in tanRouter thou

**Reply by maelp** [reply]
Still not working on Tanstack Router

**Reply by divyeshio** [reply]
@maelp Try this :

```
export const Route = createFileRoute("/")({
  component: Home,
  loader: async () => await getCount(),
  head: () => ({
    meta: [
      {
        title: "This is working!",
      },
    ],
  }),
});
```

**Reply by maelp** [reply]
@divyeshio indeed this works, I guess the doc needs to be updated to use this pattern

**Comment by nanaaikinson** [comment]
This is what I currently use and works well. You can add more features to it.

```tsx
import { useEffect } from "react";

interface HelmetProps {
  title?: string;
  description?: string;
  keywords?: string;
}

const Helmet: React.FC<HelmetProps> = ({ title, description, keywords }) => {
  useEffect(() => {
    // Set document title if provided
    if (title) {
      document.title = title;
    }

    // Set meta description if provided
    if (description) {
      let metaDescription = document.querySelector('meta[name="description"]');

      if (!metaDescription) {
        metaDescription = document.createElement("meta");

        (metaDescription as HTMLMetaElement).name = "description";

        document.head.appendChild(metaDescription);
      }
      (metaDescription as HTMLMetaElement).content = description;
    }

    // Set meta keywords if provided
    if (keywords) {
      let metaKeywords = document.querySelector('meta[name="keywords"]');

      if (!metaKeywords) {
        metaKeywords = document.createElement("meta");

        (metaKeywords as HTMLMetaElement).name = "keywords";

        document.head.appendChild(metaKeywords);
      }
      (metaKeywords as HTMLMetaElement).content = keywords;
    }
  }, [title, description, keywords]);

  return null; // This component does not render any visible UI
};

export { Helmet };

```

**Comment by BrendanC23** [comment]
The [docs](https://tanstack.com/router/latest/docs/framework/react/guide/document-head-management) are slightly outdated currently, but they describe a new way to set the `title` tag (that works with or without Start). They currently suggest doing

```tsx
export const Route = createRootRoute()({
  head: () => ({
    title: 'My App',
})
```

but the actual syntax is

```tsx
export const Route = createRootRoute()({
    head: (ctx) => {
        return {
            meta: [
                {
                    title: "My App",
                },
            ],
        };
    },
})
```

You also need to render a `<HeadContent />` tag. If you are using TanStack Start, this goes inside `<head>`. If not, it should just go before `<Outlet>`. Note that for the latter, I also had to get rid of `<title>` tag within my `<head>` in `index.html`. Otherwise, that was used instead of the route's title.

**Reply by dohomi** [reply]
I am using it with the router and this is the output of my HTML based on follwoing __root.tsx. Shouldnt the content be part of `<head>` instead of `<body>`?

<img width="1102" alt="Screenshot 2025-02-21 at 10 22 45" src="https://github.com/user-attachments/assets/03dfef92-10d4-49e3-9526-bfa8d2c01fd2" />
<img width="356" alt="Screenshot 2025-02-21 at 10 21 47" src="https://github.com/user-attachments/assets/20ac94b9-c3d9-4278-841f-365a7efe8551" />

**Reply by abrahimzaman360** [reply]
Look at my answer below:

**Reply by rhuanbarreto** [reply]
@tannerlinsley I added a PR for adding this important info for SPAs: https://github.com/TanStack/router/pull/4457

**Comment by abrahimzaman360** [comment]
The way I have done suggested by @BrendanC23:

- task:
Remove <title> tag from Index.html

• Do it once, In the Root Route file:
```js
export const Route = createRootRoute({
  component: () => (
    <>
      <HeadContent />
      <RootComponent />
    </>
  ),
  head: () => ({
    meta: [
      {
        title: "My App",
      },
      { name: "description", content: "Learn more about MyApp" },
    ],
  }),
});
```

then,
• In File Route or Virtual Route:
```js
import { createFileRoute } from "@tanstack/react-router";

export const Route = createFileRoute("/")({
  component: HomeComponent,
  head: (ctx) => {
    return {
      meta: [
        {
          title: "Index Page",
        },
        {
          name: "description",
          content: "Super Content",
        },
      ],
    };
  },
});

function HomeComponent() {
  return <h1>Hello Index</h1>;
}
```

**Reply by henggana** [reply]
Nice! `ctx` have `loaderData` property which contains data returned from the route loader.

**Reply by SolaymanSinha** [reply]
I'm wondering, how can I add dynamic meta tags? I am fetching data from my backend, and I wanna use that data to generate meta tags.

**Comment by alexy4744** [comment]
For SPAs, you can render the `<HeadContent />` inside the `<head>` of the document by using a portal, so that it isn't rendered inside of your `<body>`:

```tsx
import { createPortal } from "react-dom";

export const Route = createRootRoute({
  component: () => (
    <>
      {createPortal(<HeadContent />, document.querySelector("head")!)}

      <Outlet />
    </>
  ),
  head: () => ({
    meta: [
      { title: "My App" },
      { name: "description", content: "Learn more about MyApp" },
    ],
  }),
});
```

**Reply by dohomi** [reply]
Is this the recommended way? Isnt that something the HeadContent component might should solve internally?

**Reply by alexy4744** [reply]
Looking at the source code for HeadContent and Asset, it seems like the object returned in the `head()` function simply gets mapped to the appropriate `<title>`, `<meta>`, etc... tags. So I don't see why portal would be a bad solution. But yeah I do agree with you that this is something that should be done by the tanstack router instead.

**Reply by tannerlinsley** [reply]
Given that spas already are too late to render into head directly, there’s no incentive to force them into head other than semantic correctness. Like you mentioned, this can be done with a portal if you wish, but the runtime result would be the same if you didn’t. I could be swayed to include something like this in the core, but not easily given how simple it is to do in user land.

**Reply by maxime-forveille** [reply]
The benefit would be to take advantage of the dedup.
Right now, either you don't put the title inside the `index.html` but you will have those default route path titles or you put the title inside the `<body>`  in your `index.html` and use the portal to allow us to override it inside routes.

Example of the lack of a title in the `index.html`, we should only see `Admin App` then `Liste des organisations`:

https://github.com/user-attachments/assets/a1cfab40-d270-4612-9bd7-b6b2dff1f545

**Reply by tannerlinsley** [reply]
When would you not have the flicker though (even if it’s empty, you’ll see a flicker from nothing or the default title)? It’s too late in the document lifecycle no matter how you slice it. Not sure what else could be done here other than full document control + SSR

**Reply by maxime-forveille** [reply]
I was assuming that if my `index.html` has a defaut title in the `head` tag and I set the same at the index route, there will be no flicker.

**Reply by cfmano** [reply]
The problem with this approach is that if you only use portal for SPAs, the `HeadContent` component doesn't update existing title (and other head related elements), but only adds them into head. If you need default title (inside e.g. index.html), this approach will never work. You either need to remove original `<title>` manually or fully replace `<head>` with `<HeadContent />`. Neither of them is comfortable.

**Reply by maxime-forveille** [reply]
I understand that you don't want to parse the `index.html` in SPA mode. But the documentation could state it in a clearer way.
First I was expecting a behavior like `react-helmet-async` where I can put a `title` inside my `index.html` and then overwrite it in components.

**Comment by binhonglee** [comment]
Not an answer (more like a followup question), is there a way to dynamically set the title based on an async request response for SSR? Currently it seems to be possible to set html tags based per route but I'd like to send a separate network request to (another lol) server to get information to render the body (which can be done with `use()`) as well as `<title />` and other `meta` tags. Basically, for SEO and link preview purposes, I'd like the page to render dynamically with info received from another server.

**Reply by binhonglee** [reply]
btw i figured this out a while ago. I did something like this for my use case. Hope this helps!

```ts
export const Route = createFileRoute('/obj/view/$id')({
  loader: async ({ params: { id }, context }) => {
    return await context.queryClient.ensureQueryData(
      objQueryOptions(id),
    )
  },
  head: ({ loaderData }) => ({
    meta: seo({
      title: loaderData.name,
      description: loaderData.description,
  }),
  component: YourComponent,
})

function YourComponent() {
  ...
}
```

with `objQueryOptions` living in a separate util file (using `@tanstack/react-query`) something like this:

```ts
export const fetchObj = createServerFn({ method: 'GET' })
  .validator((d: string) => d)
  .handler(async ({ data }) => {
    return await axios
      .get<Obj>(`https://yourwebsite.com/api/${data}`)
      .then((r) => r.data)
      .catch((err) => {
        console.error(err)
        if (err.status === 404) {
          throw notFound()
        }
        throw err
      })
  })

export const objQueryOptions = (id: string) =>
  queryOptions({
    queryKey: ['post', id],
    queryFn: () => fetchObj({ data: id }),
  })
```

**Comment by amalkurbanbaev** [comment]
I couldn't find a solution that I could use without "as unknown as"

```typescript
export const Route = createFileRoute("/_protected/publications/$id/")({
  component: PublicationView,
  head: (ctx) => {
    const publication =
      ctx.loaderData as unknown as PublicationRetrieveSchemaReadable

    return {
      meta: [
        {
          title: `Publications / ${publication?.title}`,
        },
      ],
    }
  },
  loader: async ({ params, context }) => {
    const publication = await context.queryClient.ensureQueryData(
      publicationApiGetPublicationOptions({
        path: { publication_id: params.id },
      }),
    )

    return publication
  },
})
```

**Reply by schiller-manuel** [reply]
your property order is incorrect. see https://tanstack.com/router/latest/docs/eslint/create-route-property-order

**Reply by amalkurbanbaev** [reply]
wow, I didn't know about this ordering rule (I use Biome)... now the type inference works and i removed "as unknown as"

**Comment by chanithecat** [comment]
I found a way to achieve this by using `staticData`

i set an optional static data `title` to the module declaration
```ts
declare module "@tanstack/react-router" {
	interface Register {
		router: ReturnType<typeof createRouter>;
	}
  interface StaticDataRouteOption {
    title?: string
  }
}
```

then we add this to the `__root.tsx`'s title
```ts
export const Route = createRootRoute({
	head: (ctx) => ({
		meta: [
			{// other meta tags... }
			{
				title: ctx.matches[ctx.matches.length - 1].staticData.title
					? `${import.meta.env.VITE_APP_NAME} - ${ctx.matches[ctx.matches.length - 1].staticData.title}`
					: `${import.meta.env.VITE_APP_NAME}`,
			},
		],
		{// other head tags... }
	}),
	{// other options ... }
});
```
this checks if the deepest matched route has a `title` static data
if it exists we set it like `My App - Title`
else its just `My App`

then we just set the title to the child routes
```ts
// signup.tsx
export const Route = createFileRoute("/signup")({
	staticData: {
		title: "Sign Up",
	},
	component: SignupPage,
});
``` 

we get this
<img width="285" height="104" alt="image" src="https://github.com/user-attachments/assets/318d94ed-aaf0-4b03-905e-98adba5e4082" />


added bonus if i dont want the app name in there i could just simply override this by adding a head to the child as well like this
```ts
export const Route = createFileRoute("/demo/start/api-request")({
	head: () => ({
		meta: [
			{
				title: "API Request Example",
			},
		],
	}),
	component: Home,
});
``` 
<img width="282" height="140" alt="image" src="https://github.com/user-attachments/assets/375a889c-4f33-407e-92cc-cc0022c9b232" />


so far its working for me but i havent tested this in all cases.
is this correct guys? what do u guys think did i miss anything?

edit: added images

**Reply by DamianoMagrini** [reply]
Using `staticData` here limits this approach to the cases when the title is available synchronously. If a route's title relies on its loader data or other async ops, `context` can be used instead and should cover all cases.

**Edit:** Actually, while this is possible, the `context` function does not get access to `loaderData`, so e.g. database queries would need to be run twice, which is not ideal.

**Comment by maelp** [comment]
I'm trying to follow https://tanstack.com/router/latest/docs/framework/react/guide/document-head-management but setting the head({ title}) in a route does not seem to work... I've included <HeadContent> and <Scripts>, not sure what I'm missing?

**Reply by rhuanbarreto** [reply]
From docs:
> First, remove the <title> tag from the the index.html if you have set any.

Did you do this?

**Reply by michalmilek** [reply]
> From docs:
> 
> > First, remove the <title> tag from the the index.html if you have set any.
> 
> Did you do this?

no need propably(?), i had the same problem - you have to add HeadContent from tanstack router into RouterProvider

**Comment by mirus-ua** [comment]
Hello
On the docs page https://tanstack.com/router/latest/docs/framework/react/guide/document-head-management#deduping I see a paragraph with this text
> title tags defined in nested routes will override a title tag defined in a parent route (but you can compose them together, which is covered in a future section of this guide)

But not in that article or in the documentation, I can't find how to compose titles. Can you point me in the right direction? Or is this part not ready yet?

**Reply by maelp** [reply]
you can override, but the field given in the docs (at least when I checked) was the wrong one, the correct one is as above, eg

```ts
export const Route = createFileRoute("/demo/start/api-request")({
	head: () => ({
		meta: [
			{
				title: "API Request Example",
			},
		],
	}),
	component: Home,
});```

**Reply by mirus-ua** [reply]
Yes, but I want to compose the parent and child titles and not override

**Reply by divyeshio** [reply]
@mirus-ua I suppose you may want to do this for breadcrumbs. Here's how you can do it:
```
const matches = useMatches();

  const breadcrumbs = matches
    .map((m) => ({
      title: m.meta?.at(0)?.name ?? m.meta?.at(0)?.title,
      link: m.pathname,
    }))
    .filter((x) => x.title);
```

**Reply by maelp** [reply]
@divyeshio where would you put this code? in the root page? or in each subpage?

**Reply by divyeshio** [reply]
This goes into root

**Reply by mirus-ua** [reply]
@divyeshio, no. I want it for the title itself because my page structure is pretty flat. 

I want to specificly use a first-class solution that is described on the page that I linked

**Reply by strawberrymetropolis** [reply]
> Yes, but I want to compose the parent and child titles and not override

@mirus-ua , have you figured out how to do it?

**Reply by mirus-ua** [reply]
> > Yes, but I want to compose the parent and child titles and not override
> 
> @mirus-ua , have you figured out how to do it?

Unfortunately, no. I guess this isn't implemented yet

**Comment by VMBindraban** [comment]
They have updated the documentation for setting the page title and other data in the header.

https://tanstack.com/router/latest/docs/framework/react/guide/document-head-management

```js
export const Route = createFileRoute("/some-route")({
  head: () => ({
    meta: [
      {
        title: 'My App',
      },
    ],
  )
);
```

You have to set `<HeadContent />` as described at https://tanstack.com/router/latest/docs/framework/react/guide/document-head-management#single-page-applications.

**Reply by everdrone** [reply]
Still, there is no way of composing titles from child routes like `<child title> | <root title>`?

**Reply by IanVS** [reply]
I'm looking for this as well.  The docs make it sound like it's just not yet documented, but maybe it's just not built yet?

**Comment by ricsands2801** [comment]
@tannerlinsley is there a proper way to set dynamic titles e.g. $orderId.tsx in a single page app sets the order ID in the title after it is loading from tanstack query?
